rasa/shared/core/domain.py

```diff
@@ -1590,6 +1596,9 @@ def is_domain_file(filename: Text) -> bool:
         except (ValueError, YamlSyntaxException):
             return False
 
+        if any(key in content for key in DISALLOWED_DOMAIN_KEYS):
```

Comment:
isn't this quite dangerous? What if want to put everything in one file, e.g. mix stories with the domain file? That should in my opinion still work.

Reply:
I was basing it off of the docs [here](https://rasa.com/docs/rasa/domain) where we have a full example of a domain file which doesn't include any of the ones that are not allowed in this PR.

Reply:
Assuming that the keys that are not listed on that page are generally not allowed in a domain file, is that assumption not correct?

Reply:
Afaik everything can be merged / split with the Rasa Open Source 2.0 YAML format as you want (cc @degiz ). For me the problem is how Rasa X handles these new yaml format capabilities and that we haven't resolved this properly. Not a huge fan of fixing things in Rasa Open Source for that.

Reply:
> everything can be merged / split with the Rasa Open Source 2.0 YAML 

not really, domain still stays aside. The only thing from `domain` that's shared with `nlu` are the `responses`.

Reply:
We clearly made this distinction, domain stays separate and **has** to be specified with a `-d` flag with all rasa cli related commands.

Reply:
More generally, we have:
- training data files (nlu, stories, rules, responses)
- domain files
- configuration files (endpoints, config, credentials)

Reply:
> We clearly made this distinction

Thanks for clarifying 🤔 Did we write that down somewhere?

What about the server endpoints which accept YAML in Rasa? In that case all data is put in one big payload, no? Btw that was [my PR](https://github.com/RasaHQ/rasa/pull/6372/files) back then and some related discussion [here](https://github.com/RasaHQ/rasa/issues/6354#issuecomment-671328929) (Tom intervened to explicitly not split things 😄 ). 

If Rasa X can't deal with mixed files and Rasa Open Source currently can - shouldn't we rather add support for this to Rasa X then?

Reply:
Also in the [docs here](https://rasa.com/docs/rasa/chitchat-faqs#5-defining-the-responses) we explictly recommend putting responses in the domain file 😬

Reply:
> shouldn't we rather add support for this to Rasa X then?

That's the next step (see [this scope doc](https://www.notion.so/rasa/Rasa-X-dumping-mixed-training-files-d02e7973f88942ec912583f5bd1aa29a)).

But even if Rasa X is able to write mixed files, I don't think there's a way for Rasa X to distinguish between a domain file and an NLU file with responses at the moment, which means that such a file would show up on the domain screen.

Reply:
But shouldn't these responses show up in the domain screen? Just the other NLU stuff shouldn't show up, no?

Reply:
> But shouldn't these responses show up in the domain screen? Just the other NLU stuff shouldn't show up, no?

I guess that's another question. The responses from NLU files should show up in the Responses screen, not sure if I have an answer if they should show up in the domain screen.

Reply:
> Also in the docs here we explictly recommend putting responses in the domain file 😬

Another possible solution for this problem ultimately [this bug](https://github.com/RasaHQ/rasa-x/issues/4252) could also be to not allow responses in NLU files anymore.
I think what's missing at the moment is a clear direction of what kind of data is allowed in which files, or even a documented decision stating that a user can mix and match to their heart's content.

Reply:
> could also be to not allow responses in NLU files anymore.

This seems all a bit fragile to me to be honest. What is the current problem with files with `responses` being recognized as domain files? Is it because _1_ file is referenced by _two_ Rasa X entity objects which then override each other during the dump?

Reply:
I think we're honing in too fast on the "X overrides mixed data" problem, which in my opinion is a separate issue and wouldn't even solve [this bug](https://github.com/RasaHQ/rasa-x/issues/4252).

The more interesting issue to discuss about is rather that:
* X (currently) needs to differentiate between domain and NLU files (amongst others), and it needs to do that since we differentiate between a mix of files and object types in the UI
  *  i.e. Domain screen is for the domain file
  * Responses, "NLU data", Responses, etc. screens are around object types
* An NLU file that has `responses` should not be identified as a domain file

The overwriting of the file content is rather a symptom of the assumption that any file with at least one of the allowed domain file keys is in fact a domain file, which I think is an assumption that is worth challenging from a product perspective.